// @polymeshassociation/polymesh-sdk/api/procedures/clearMetadata.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearMetadata = exports.getAuthorization = exports.prepareClearMetadata = void 0;
const internal_1 = require("../../internal");
const types_1 = require("../../types");
const conversion_1 = require("../../utils/conversion");
/**
 * @hidden
 */
function prepareClearMetadata(params) {
    return __awaiter(this, void 0, void 0, function* () {
        const { context: { polymeshApi: { tx }, }, context, } = this;
        const { metadataEntry: { id, type, asset: { ticker }, }, metadataEntry, } = params;
        const rawTicker = (0, conversion_1.stringToTicker)(ticker, context);
        const rawMetadataKey = (0, conversion_1.metadataToMeshMetadataKey)(type, id, context);
        const { canModify, reason } = yield metadataEntry.isModifiable();
        if (!canModify) {
            throw reason;
        }
        return {
            transaction: tx.asset.removeMetadataValue,
            args: [rawTicker, rawMetadataKey],
            resolver: undefined,
        };
    });
}
exports.prepareClearMetadata = prepareClearMetadata;
/**
 * @hidden
 */
function getAuthorization(params) {
    const { context } = this;
    const { metadataEntry: { asset: { ticker }, }, } = params;
    return {
        permissions: {
            transactions: [types_1.TxTags.asset.RemoveMetadataValue],
            assets: [new internal_1.FungibleAsset({ ticker }, context)],
            portfolios: [],
        },
    };
}
exports.getAuthorization = getAuthorization;
/**
 * @hidden
 */
const clearMetadata = () => new internal_1.Procedure(prepareClearMetadata, getAuthorization);
exports.clearMetadata = clearMetadata;
